src/proposals.rs

```diff
@@ -0,0 +1,269 @@
+                match client.get_latest_block(GetLatestBlockRequest {}).await {
+                    Ok(r) => {
+                        let r = r.into_inner();
+                        if r.block.is_some() {
+                            let block = r.block.unwrap();
+                            if block.header.is_some() {
+                                self.last_observed_height = block.header.unwrap().height as u64;
+                            }
+                        }
+                    }
+                    Err(err) => warn!("failed to query latest block height. this could eventually result in degraded scheduled cork proposal query performance: {}", err),
+                };
+            }
+            Err(err) => warn!("failed to query latest block height. this could eventually result in degraded scheduled cork proposal query performance: {}", err),
+        };
+    }
+
+    fn increment_proposal_id(&mut self) {
+        self.last_processed_proposal_id += 1;
+    }
+}
+
+async fn poll_approved_proposals(
+    state: &mut ProposalThreadState,
+    publisher_cache_tx: Sender<()>,
+) -> Result<(), Error> {
+    let config = APP.config();
+    let mut gov_client = GovQueryClient::connect(config.cosmos.grpc.clone()).await?;
+
+    loop {
+        // Proposal IDs start at 1, so this should be ok even for the first query after startup.
+        let proposal_id = state.last_processed_proposal_id + 1;
+        debug!("querying proposal {}", proposal_id);
+        let proposal = match gov_client
+            .proposal(tonic::Request::new(QueryProposalRequest { proposal_id }))
+            .await
+        {
+            Ok(r) => r.into_inner().proposal,
+            Err(err) => {
+                if err.code() == Code::NotFound {
+                    // look ahead in case id was skipped
+                    let mut missing_proposals = 1;
+                    let mut found_proposal = false;
+
+                    for i in 1..=10 {
+                        if let Err(err) = gov_client
+                            .proposal(tonic::Request::new(QueryProposalRequest {
+                                proposal_id: proposal_id + i,
+                            }))
+                            .await
+                        {
+                            if err.code() == Code::NotFound {
+                                missing_proposals += 1;
+                                continue;
+                            } else {
+                                return Err(proposal_processing_error(format!(
+                                    "error querying proposal {}: {}",
+                                    proposal_id, err
+                                )));
+                            }
+                        }
+
+                        found_proposal = true;
+                        break;
+                    }
+
+                    if found_proposal {
+                        state.last_processed_proposal_id += missing_proposals;
```

Comment:
We're counting missing proposals, but when we find one, won't that have the ID? Couldn't we read it directly?

Reply:
It's not actually processing the found one, it's setting the ID to the one right before the one that was found so that the next loop can process it.